File: web/src/lib/scoreScale.ts
/**
 * Generic A–F score-scale helpers shared across pages that render a
 * letter-grade badge (Drives, Charging, future Trips). Pure functions
 * with no React or domain knowledge so any caller can use them.
 *
 * The palette and thresholds match the existing per-drive grade badge
 * (see `lib/drivesAggregation.ts > GRADE_PALETTE`) so a screen showing
 * both a Drive grade and a Charging grade uses the same colours for
 * "A", "B", etc — no per-page palette drift.
 */

export type ScoreGrade = 'A+' | 'A' | 'B' | 'C' | 'D' | 'F' | '—';

export interface ScoreGradeInfo {
  /** Display label. `—` when input is null / NaN. */
  label: ScoreGrade;
  /** Hex colour for the badge text. */
  color: string;
  /**
   * Numeric weight for averaging across many items. `null` for the "no
   * data" sentinel so callers can skip it in arithmetic.
   */
  numeric: number | null;
}

/** Shared palette. Keep in lock-step with `lib/drivesAggregation` GRADE_PALETTE. */
const GRADE_PALETTE: Record<ScoreGrade, { color: string; numeric: number | null }> = {
  'A+': { color: '#10b981', numeric: 4.5 },
  A:    { color: '#10b981', numeric: 4.0 },
  B:    { color: '#00f0ff', numeric: 3.0 },
  C:    { color: '#f59e0b', numeric: 2.0 },
  D:    { color: '#ef4444', numeric: 1.0 },
  F:    { color: '#b91c1c', numeric: 0.5 },
  '—':  { color: '#6b7280', numeric: null },
};

/** Default 0–100 thresholds. Lower bound inclusive. */
export const DEFAULT_SCORE_THRESHOLDS: ReadonlyArray<{ min: number; label: ScoreGrade }> = [
  { min: 90, label: 'A+' },
  { min: 80, label: 'A'  },
  { min: 65, label: 'B'  },
  { min: 50, label: 'C'  },
  { min: 35, label: 'D'  },
  { min: 0,  label: 'F'  },
];

/**
 * Map a 0–100 numeric score to a letter grade. Caller can override
 * thresholds (Wh/km efficiency, latency ms, anything ordered).
 */
export function numericToGrade(
  score: number | null | undefined,
  thresholds: ReadonlyArray<{ min: number; label: ScoreGrade }> = DEFAULT_SCORE_THRESHOLDS,
): ScoreGradeInfo {
  if (score == null || !Number.isFinite(score)) {
    return { label: '—', ...GRADE_PALETTE['—'] };
  }
  // Thresholds are evaluated highest-first so the first match wins.
  const sorted = [...thresholds].sort((a, b) => b.min - a.min);
  for (const t of sorted) {
    if (score >= t.min) {
      return { label: t.label, ...GRADE_PALETTE[t.label] };
    }
  }
  return { label: 'F', ...GRADE_PALETTE.F };
}

/** Lookup the colour + numeric weight for a known grade label. */
export function gradeInfo(label: ScoreGrade): ScoreGradeInfo {
  return { label, ...GRADE_PALETTE[label] };
}

/**
 * Average a list of grade-numerics (skipping `null`) and map back to
 * a letter grade. Returns `—` when no graded inputs are present.
 */
export function averageGrade(values: ReadonlyArray<number | null>): ScoreGradeInfo {
  let sum = 0;
  let n = 0;
  for (const v of values) {
    if (v != null && Number.isFinite(v)) {
      sum += v;
      n += 1;
    }
  }
  if (n === 0) return { label: '—', ...GRADE_PALETTE['—'] };
  // Map averaged numeric back to a label using inverse of GRADE_PALETTE.
  const avg = sum / n;
  if (avg >= 4.25) return { label: 'A+', ...GRADE_PALETTE['A+'] };
  if (avg >= 3.5)  return { label: 'A',  ...GRADE_PALETTE.A };
  if (avg >= 2.5)  return { label: 'B',  ...GRADE_PALETTE.B };
  if (avg >= 1.5)  return { label: 'C',  ...GRADE_PALETTE.C };
  if (avg >= 0.75) return { label: 'D',  ...GRADE_PALETTE.D };
  return { label: 'F', ...GRADE_PALETTE.F };
}

File: apps/windows/TeslaSync.App.Core/DataDisplay/ScoreScale.cs
using System.Collections.Generic;

namespace TeslaSync.App.Core.DataDisplay;

/// <summary>Letter grade rendered by <c>TsScoreBadge</c> (A+ … F, or — for no data).</summary>
public enum ScoreGrade
{
    APlus,
    A,
    B,
    C,
    D,
    F,
    None,
}

/// <summary>A score grade together with its palette colour and numeric weight.</summary>
/// <param name="Grade">The letter grade.</param>
/// <param name="Label">Display label ("A+", "A", …, "—").</param>
/// <param name="ColorHex">Hex colour for the badge text.</param>
/// <param name="Numeric">Numeric weight for averaging; null for the no-data sentinel.</param>
public readonly record struct ScoreGradeInfo(ScoreGrade Grade, string Label, string ColorHex, double? Numeric);

/// <summary>A score threshold: scores at or above <see cref="Min"/> map to <see cref="Grade"/>.</summary>
public readonly record struct ScoreThreshold(double Min, ScoreGrade Grade);

/// <summary>
/// Generic A–F score-scale helpers (port of web <c>scoreScale.ts</c>). Pure
/// functions with no UI knowledge. The palette and thresholds match the web
/// source so a screen showing both a Drive grade and a Charging grade uses the
/// same colours for "A", "B", etc.
/// </summary>
public static class ScoreScale
{
    private static readonly Dictionary<ScoreGrade, ScoreGradeInfo> Palette =
        new()
        {
            [ScoreGrade.APlus] = new(ScoreGrade.APlus, "A+", "#10b981", 4.5),
            [ScoreGrade.A] = new(ScoreGrade.A, "A", "#10b981", 4.0),
            [ScoreGrade.B] = new(ScoreGrade.B, "B", "#00f0ff", 3.0),
            [ScoreGrade.C] = new(ScoreGrade.C, "C", "#f59e0b", 2.0),
            [ScoreGrade.D] = new(ScoreGrade.D, "D", "#ef4444", 1.0),
            [ScoreGrade.F] = new(ScoreGrade.F, "F", "#b91c1c", 0.5),
            [ScoreGrade.None] = new(ScoreGrade.None, "\u2014", "#6b7280", null),
        };

    /// <summary>Default 0–100 thresholds (lower bound inclusive), highest first.</summary>
    public static IReadOnlyList<ScoreThreshold> DefaultThresholds { get; } = new[]
    {
        new ScoreThreshold(90, ScoreGrade.APlus),
        new ScoreThreshold(80, ScoreGrade.A),
        new ScoreThreshold(65, ScoreGrade.B),
        new ScoreThreshold(50, ScoreGrade.C),
        new ScoreThreshold(35, ScoreGrade.D),
        new ScoreThreshold(0, ScoreGrade.F),
    };

    /// <summary>Map a 0–100 numeric score to a letter grade.</summary>
    public static ScoreGradeInfo NumericToGrade(double? score, IReadOnlyList<ScoreThreshold>? thresholds = null)
    {
        if (score is not { } s || double.IsNaN(s) || double.IsInfinity(s))
        {
            return Palette[ScoreGrade.None];
        }

        var src = thresholds ?? DefaultThresholds;
        var sorted = new List<ScoreThreshold>(src);
        sorted.Sort((a, b) => b.Min.CompareTo(a.Min));
        foreach (var t in sorted)
        {
            if (s >= t.Min)
            {
                return Palette[t.Grade];
            }
        }

        return Palette[ScoreGrade.F];
    }

    /// <summary>Look up the colour + numeric weight for a known grade.</summary>
    public static ScoreGradeInfo Info(ScoreGrade grade) => Palette[grade];

    /// <summary>
    /// Average a list of grade-numerics (skipping null) and map back to a letter.
    /// Returns the no-data sentinel when no graded inputs are present.
    /// </summary>
    public static ScoreGradeInfo AverageGrade(IEnumerable<double?> values)
    {
        double sum = 0;
        int n = 0;
        foreach (var v in values)
        {
            if (v is { } d && !double.IsNaN(d) && !double.IsInfinity(d))
            {
                sum += d;
                n++;
            }
        }

        if (n == 0)
        {
            return Palette[ScoreGrade.None];
        }

        double avg = sum / n;
        if (avg >= 4.25)
        {
            return Palette[ScoreGrade.APlus];
        }

        if (avg >= 3.5)
        {
            return Palette[ScoreGrade.A];
        }

        if (avg >= 2.5)
        {
            return Palette[ScoreGrade.B];
        }

        if (avg >= 1.5)
        {
            return Palette[ScoreGrade.C];
        }

        if (avg >= 0.75)
        {
            return Palette[ScoreGrade.D];
        }

        return Palette[ScoreGrade.F];
    }
}
